Build failure on branch-2.2.x

Commit: a2df0c035e9e49fbcdb7644b0079930db5155800

When building with java8, it ran into the build error:
```
$ java -version
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_265-b01)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.265-b01, mixed mode)

$ ./mvnw -P hadoop3 clean install -DskipTests
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project gcsio: Compilation failure: Compilation failure: 
[ERROR] /home/veblush/hadoop-connectors/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:[22,44] cannot find symbol
[ERROR]   symbol:   class GoogleCloudStorageStatistics
[ERROR]   location: package com.google.cloud.hadoop.gcsio
[ERROR] /home/veblush/hadoop-connectors/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:[22,1] static import only from classes and interfaces
[ERROR] /home/veblush/hadoop-connectors/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:[156,19] cannot find symbol
[ERROR]   symbol:   class GoogleCloudStorageStatistics
[ERROR]   location: class com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl
[ERROR] /home/veblush/hadoop-connectors/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:[2411,41] cannot find symbol
[ERROR]   symbol:   class GoogleCloudStorageStatistics
[ERROR]   location: class com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl
```

#653 looks relevant to this error.

--- a/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
+++ b/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java
@@ -19,7 +19,6 @@ package com.google.cloud.hadoop.gcsio;
 import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createFileNotFoundException;
 import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createJsonResponseException;
 import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageItemInfo.createInferredDirectory;
-import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageStatistics.OBJECT_DELETE_OBJECTS;
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
@@ -58,6 +57,7 @@ import com.google.api.services.storage.model.Objects;
 import com.google.api.services.storage.model.RewriteResponse;
 import com.google.api.services.storage.model.StorageObject;
 import com.google.auth.Credentials;
+import com.google.cloud.hadoop.gcsio.authorization.StorageRequestAuthorizer;
 import com.google.cloud.hadoop.util.AccessBoundary;
 import com.google.cloud.hadoop.util.ApiErrorExtractor;
 import com.google.cloud.hadoop.util.BaseAbstractGoogleAsyncWriteChannel;
@@ -104,7 +104,6 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 import javax.annotation.Nullable;
 
@@ -151,10 +150,6 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
           "metadata");
 
   private static final String LIST_OBJECT_FIELDS_FORMAT = "items(%s),prefixes,nextPageToken";
-
-  // To track the object statistics
-  private HashMap<GoogleCloudStorageStatistics, AtomicLong> objectStatistics =
-      new HashMap<GoogleCloudStorageStatistics, AtomicLong>();
 
   // A function to encode metadata map values
   static String encodeMetadataValues(byte[] bytes) {
@@ -265,6 +260,9 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
 
   // Determine if a given IOException is due to rate-limiting.
   private RetryDeterminer<IOException> rateLimitedRetryDeterminer = errorExtractor::rateLimited;
+
+  // Authorization Handler instance.
+  private final StorageRequestAuthorizer storageRequestAuthorizer;
 
   // Function that generates downscoped access token.
   private final Function<List<AccessBoundary>, String> downscopedAccessTokenFn;
@@ -387,6 +385,7 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
       }
     }
 
+    this.storageRequestAuthorizer = initializeStorageRequestAuthorizer(storageOptions);
     this.downscopedAccessTokenFn = downscopedAccessTokenFn;
   }
 
@@ -395,12 +394,25 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
       throws IOException {
     HttpTransport httpTransport =
         HttpTransportFactory.createHttpTransport(
-            options.getProxyAddress(), options.getProxyUsername(), options.getProxyPassword());
+            options.getTransportType(),
+            options.getProxyAddress(),
+            options.getProxyUsername(),
+            options.getProxyPassword());
     return new Storage.Builder(httpTransport, JSON_FACTORY, httpRequestInitializer)
         .setRootUrl(options.getStorageRootUrl())
         .setServicePath(options.getStorageServicePath())
         .setApplicationName(options.getAppName())
         .build();
+  }
+
+  @VisibleForTesting
+  static StorageRequestAuthorizer initializeStorageRequestAuthorizer(
+      GoogleCloudStorageOptions options) {
+    return options.getAuthorizationHandlerImplClass() == null
+        ? null
+        : new StorageRequestAuthorizer(
+            options.getAuthorizationHandlerImplClass(),
+            options.getAuthorizationHandlerProperties());
   }
 
   private ExecutorService createManualBatchingThreadPool() {
@@ -734,48 +746,25 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
     Preconditions.checkArgument(
         resourceId.isStorageObject(), "Expected full StorageObject id, got %s", resourceId);
 
+    if (storageOptions.isGrpcEnabled()) {
+      return GoogleCloudStorageGrpcReadChannel.open(
+          storageStubProvider, storage, errorExtractor, resourceId, readOptions);
+    }
+
     // The underlying channel doesn't initially read data, which means that we won't see a
     // FileNotFoundException until read is called. As a result, in order to find out if the
-    // object exists, we'll need to do an RPC (metadata or data). A metadata check should be a less
+    // object
+    // exists, we'll need to do an RPC (metadata or data). A metadata check should be a less
     // expensive operation than a read data operation.
-    GoogleCloudStorageItemInfo itemInfo =
-        readOptions.getFastFailOnNotFound() ? getItemInfo(resourceId) : null;
-
-    return open(resourceId, itemInfo, readOptions);
-  }
-
-  /**
-   * See {@link GoogleCloudStorage#open(GoogleCloudStorageItemInfo)} for details about expected
-   * behavior.
-   */
-  public SeekableByteChannel open(
-      GoogleCloudStorageItemInfo itemInfo, GoogleCloudStorageReadOptions readOptions)
-      throws IOException {
-    logger.atFiner().log("open(%s, %s)", itemInfo, readOptions);
-    checkNotNull(itemInfo, "itemInfo should not be null");
-
-    StorageResourceId resourceId = itemInfo.getResourceId();
-    Preconditions.checkArgument(
-        resourceId.isStorageObject(), "Expected full StorageObject id, got %s", resourceId);
-
-    return open(resourceId, itemInfo, readOptions);
-  }
-
-  private SeekableByteChannel open(
-      StorageResourceId resourceId,
-      GoogleCloudStorageItemInfo itemInfo,
-      GoogleCloudStorageReadOptions readOptions)
-      throws IOException {
-    if (itemInfo != null && !itemInfo.exists()) {
-      throw createFileNotFoundException(
-          resourceId.getBucketName(), resourceId.getObjectName(), /* cause= */ null);
-    }
-    if (storageOptions.isGrpcEnabled()) {
-      return itemInfo == null
-          ? GoogleCloudStorageGrpcReadChannel.open(
-              storageStubProvider, storage, errorExtractor, resourceId, readOptions)
-          : GoogleCloudStorageGrpcReadChannel.open(
-              storageStubProvider, storage, itemInfo, readOptions);
+    GoogleCloudStorageItemInfo info;
+    if (readOptions.getFastFailOnNotFound()) {
+      info = getItemInfo(resourceId);
+      if (!info.exists()) {
+        throw createFileNotFoundException(
+            resourceId.getBucketName(), resourceId.getObjectName(), /* cause= */ null);
+      }
+    } else {
+      info = null;
     }
 
     return new GoogleCloudStorageReadChannel(
@@ -784,7 +773,7 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
       @Override
       @Nullable
       protected GoogleCloudStorageItemInfo getInitialMetadata() {
-        return itemInfo;
+        return info;
       }
 
       @Override
@@ -836,17 +825,11 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
 
   public void deleteObject(StorageResourceId resourceId, long metaGeneration) throws IOException {
     String bucketName = resourceId.getBucketName();
-
     Storage.Objects.Delete deleteObject =
         initializeRequest(
                 storage.objects().delete(bucketName, resourceId.getObjectName()), bucketName)
             .setIfMetagenerationMatch(metaGeneration);
     deleteObject.execute();
-
-    //  To update the statistics of number of objects deleted
-
-    objectStatistics.putIfAbsent(OBJECT_DELETE_OBJECTS, new AtomicLong(0));
-    objectStatistics.get(OBJECT_DELETE_OBJECTS).incrementAndGet();
   }
 
   /** See {@link GoogleCloudStorage#deleteObjects(List)} for details about expected behavior. */
@@ -876,11 +859,8 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
             fullObjectNames.size(),
             storageOptions.getBatchThreads());
 
-    // update the statistics of number of objects deleted
-    objectStatistics.putIfAbsent(OBJECT_DELETE_OBJECTS, new AtomicLong(0));
     for (StorageResourceId fullObjectName : fullObjectNames) {
       queueSingleObjectDelete(fullObjectName, innerExceptions, batchHelper, 1);
-      objectStatistics.get(OBJECT_DELETE_OBJECTS).incrementAndGet();
     }
 
     batchHelper.flush();
@@ -2312,6 +2292,9 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
       String token = downscopedAccessTokenFn.apply(accessBoundaries);
       request.getRequestHeaders().setAuthorization("Bearer " + token);
     }
+    if (storageRequestAuthorizer != null) {
+      storageRequestAuthorizer.authorize(request);
+    }
     return configureRequest(request, bucketName);
   }
 
@@ -2405,10 +2388,5 @@ public class GoogleCloudStorageImpl implements GoogleCloudStorage {
     if (userProjectField != null) {
       request.set(USER_PROJECT_FIELD_NAME, projectId);
     }
-  }
-
-  @Override
-  public AtomicLong getObjectStatistics(GoogleCloudStorageStatistics key) {
-    return objectStatistics.get(key);
   }
 }

